[FIX] l10n_sa_edi: fix missing signature for POS orders

mebe-odoo opened the pull request:
When trying to print an invoice upon validation of a payment on the POS, the system returns an error, since it creates the invoice, posts it then tries to print the PDF Invoice report. At that point in time, the invoice does not have necessary data required to print the QR code on the invoice, and thus the process throws an exception.

Description of the issue/feature this PR addresses:
Creating a B2C invoice on a Saudi POS throws an exception. The POS tries to print the invoice linked to the order once it's confirmed, and the report includes a QR code that depends on data that has not been generated yet at that point in time.

Current behavior before PR:
POS throws an exception when validating a payment for a POS order linked to an individual on a Saudi database with ZATCA localization installed.

Desired behavior after PR is merged:
POS correctly validates the payment and print the report.




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/153421).

jco-odoo commented:
@mebe-odoo Won't it intervene with regular invoices if we change e.g. the address on the partner before sending?

jco-odoo commented:
@mebe-odoo Is this one still relevant?

mebe-odoo commented:
Hello @jco-odoo I updated the commit to take into account your latest comment from February 15: since we cannot accurately control when and how information on the seller/customer changes, I added a clause in `_l10n_sa_export_zatca_invoice` on line `227` where it recomputes the signature just before we sign the invoice and submit it to ZATCA.

There is a similar case to this in `l10n_mx_edi` under `_l10n_mx_edi_get_invoice_cfdi_values`

jco-odoo commented:
@mebe-odoo I still find it tricky, because we had a reason to store it in the first place: it might be in case of a timeout that we still want to verify with the original signature.  Also, maybe we could do something where we wait for the QR code to be retrieved in the PoS or not let the PDF print or even put the qr code immediately on the ticket.  (we did something with l10n_es_tbai_pos recently if you want some inspiration)

jco-odoo commented:
@mebe-odoo I am afraid this PR will mess up things with e.g. timeouts that needs to be posted again.  What do you think about merging #169407 first?

mebe-odoo commented:
> @mebe-odoo I am afraid this PR will mess up things with e.g. timeouts that needs to be posted again. What do you think about merging #169407 first?

I already left a comment on the PR you linked, the issue with passing an empty string if the signature has not been set yet is that we end up with an erroneous QR code. A better approach would be to only calculate the QR code string if the signature is set,

so basically turn this:
`        if invoice_datetime and journal_id.company_id.vat and x509_cert:`
Into this:
`        if invoice_datetime and journal_id.company_id.vat and x509_cert and signature:`

ygu-odoo commented:
@jco-odoo is this PR still valid? I checked that we did some fix in https://github.com/odoo/odoo/pull/169407 as well as https://github.com/odoo/odoo/pull/199039
